### PR TITLE
Fix the double `STEP_IN` issue for some standard libraries in macOS

### PR DESCRIPTION
**Pull request by praveennadarajah**
## Purpose
This PR will fix the double `STEP_IN` issue for some standard libraries in macOS

Fixes https://github.com/ballerina-platform/ballerina-lang/issues/29735

### CODE REVIEW

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JBallerinaDebugServer.java [hunk 1]
```diff
@@ -296,6 +297,13 @@ ClientConfigHolder getClientConfigHolder() {
                     .filter(JBallerinaDebugServer::isValidFrame)
                     .toArray(StackFrame[]::new);
 
+            if (context.getLastInstruction() == DebugInstruction.STEP_IN
+                && validFrames[0].getSource().getPath().equals(topStackFrame.getSource().getPath())
```
Comment (NipunaRanasinghe) [inline misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JBallerinaDebugServer.java:301]: Potential `IndexOutOfBoundsException` as the `validFrames` can be an empty array

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JBallerinaDebugServer.java [hunk 2]
```diff
@@ -296,6 +297,13 @@ ClientConfigHolder getClientConfigHolder() {
...
+                && validFrames[0].getLine().intValue() == topStackFrame.getLine()) {
```
Comment (NipunaRanasinghe) [inline misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JBallerinaDebugServer.java:302]: Potential `IndexOutOfBoundsException` as the `validFrames` can be an empty array

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JBallerinaDebugServer.java [hunk 3]
```diff
@@ -296,6 +297,13 @@ ClientConfigHolder getClientConfigHolder() {
...
+                eventProcessor.sendStepRequest(args.getThreadId(), StepRequest.STEP_INTO);
+                context.setLastInstruction(null);
+            }
+            topStackFrame = validFrames[0];
```
Comment (NipunaRanasinghe) [inline misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JBallerinaDebugServer.java:306]: Potential `IndexOutOfBoundsException` as the `validFrames` can be an empty array

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JBallerinaDebugServer.java [hunk 4]
```diff
@@ -296,6 +297,13 @@ ClientConfigHolder getClientConfigHolder() {
                     .filter(JBallerinaDebugServer::isValidFrame)
                     .toArray(StackFrame[]::new);
 
+            if (context.getLastInstruction() == DebugInstruction.STEP_IN
+                && validFrames[0].getSource().getPath().equals(topStackFrame.getSource().getPath())
+                && validFrames[0].getLine().intValue() == topStackFrame.getLine()) {
+                eventProcessor.sendStepRequest(args.getThreadId(), StepRequest.STEP_INTO);
+                context.setLastInstruction(null);
```
Comment (NipunaRanasinghe) [inline misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JBallerinaDebugServer.java:304]: Why are we setting `null` as the last instruction?

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JBallerinaDebugServer.java [hunk 5]
```diff
@@ -142,6 +142,7 @@
     private ThreadReferenceProxyImpl activeThread;
     private SuspendedContext suspendedContext;
     private boolean terminationRequestReceived = false;
+    private StackFrame topStackFrame;
```
Comment (NipunaRanasinghe) [inline misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JBallerinaDebugServer.java:145]: As this is basically contains the PREVIOUS breakpoint location information, shall we use a more meaningful name which indicates that?

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JBallerinaDebugServer.java (hunk 1 above)
Comment (praveennadarajah) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JBallerinaDebugServer.java]: Fixed

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JBallerinaDebugServer.java (hunk 2 above)
Comment (praveennadarajah) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JBallerinaDebugServer.java]: Fixed

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JBallerinaDebugServer.java (hunk 3 above)
Comment (praveennadarajah) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JBallerinaDebugServer.java]: Fixed

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JBallerinaDebugServer.java (hunk 4 above)
Comment (praveennadarajah) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JBallerinaDebugServer.java:305]: Fixed

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JBallerinaDebugServer.java (hunk 5 above)
Comment (praveennadarajah) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JBallerinaDebugServer.java]: Changed the name to `previousTopFrame`

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JBallerinaDebugServer.java (hunk 4 above)
Comment (praveennadarajah) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JBallerinaDebugServer.java:305]: Here I'm making sure `STEP_IN` request is sent only once, and after the request is sent I'm making the last instruction to null. 

Alternatively, we can maintain a flag to control the number of `STEP_IN` request. WDYT?

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JBallerinaDebugServer.java (hunk 5 above)
Comment (NipunaRanasinghe) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JBallerinaDebugServer.java]: Using the `topFrame` might still be misleading, when considering why we are using it. How about `lastLocationInfo`?

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JBallerinaDebugServer.java (hunk 4 above)
Comment (NipunaRanasinghe) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JBallerinaDebugServer.java:305]: i"m still not clear how we are receiving multiple step in requests from the client for a single event.

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JBallerinaDebugServer.java (hunk 5 above)
Comment (praveennadarajah) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JBallerinaDebugServer.java]: Changed to `lastDebbugHitFrame`

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JBallerinaDebugServer.java (hunk 4 above)
Comment (praveennadarajah) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JBallerinaDebugServer.java:310]: We can get multiple `stackTrace` request for a single breakpoint event. Possible `stackTrace` requests are

1.  To represent debug hit navigation bar
2. To represent call stack information.

**Review by NipunaRanasinghe — APPROVED** [review]

**Review by NipunaRanasinghe — CHANGES_REQUESTED** [review]
@praveennadarajah As per the offline discussion we had, let's move this logic to the `JDIEventProcessor` so that we can avoid additional stop event sent to the client and reduce the processing overhead.

**Review by NipunaRanasinghe — COMMENTED** [review]
According to the review we had, let's hold this until we find a proper resolution for the regression issue (related to recursive functions).

**Pull request closed without merging**